docs/fluttertodostutorial.md

```diff
@@ -11,473 +11,760 @@
+  // Persists todos to local disk and the web
+  Future saveTodos(List<TodoEntity> todos);
+}
+```
 
-Let's create a `models` directory and create `todo.dart`.
+The above is based on the implementation by [Brian Egan](https://github.com/brianegan).
+His TodosRepository is [described here](https://github.com/brianegan/flutter_architecture_samples/tree/master/todos_repository_core).
+It is shared among all of the [Todo Architecture Samples](https://github.com/brianegan/flutter_architecture_samples).
 
-[todo.dart](_snippets/flutter_todos_tutorial/todo.dart.md ':include')
+Brian Egan's `TodosRepository` has only two methods to `loadTodos` and to `saveTodos` (note the plural). Hence, a full list of Todos must be sent each time.
 
-?> **Note:** We're using the [Equatable](https://pub.dev/packages/equatable) package so that we can compare instances of `Todos` without having to manually override `==` and `hashCode`.
+- One limitation of this approach is that the standard CRUD (Create, Read, Update, and Delete) operation requires sending the full list of Todos with each repository call. For example, on an AddTodo type of screen, one cannot just send the added item. One must send the full list with the added item.
+- A second limitation is that `loadTodos` is a one-time delivery of data. `loadToDos` is not a Stream. The app must ask for updates. This is okay if an app is for a single-user, on a single device, and not connected to the cloud. However, many modern apps allow
+  - one user signed in on multiple devices at once
+  - multiple users adding to the same data repository (like a group todo-list)
 
-Next up, we need to create the `TodosState` which our presentation layer will receive.
+#### Stream approach
 
-### States
+The new way used here for `TodosRepository` has two defining differences from the original:
+1. The list of Todos is a Stream that can be subscribed to. (see `getTodos()` below)
+```dart
+/// Provides a [Stream] of all Todos.
+Stream<List<Todo>> getTodos() => _todosApi.getTodos();
+```
 
-Let's create `blocs/todos/todos_state.dart` and define the different states we'll need to handle.
+2. Individual Todos can be created, deleted, or updated. For example, both deleting and saving a Todo are done with only the `todo` as the argument. The whole list isn't needed.
+```dart
+  Future<void> saveTodo(Todo todo) => _todosApi.saveTodo(todo);
+  Future<void> deleteTodo(String id) => _todosApi.deleteTodo(id);
+```
 
-The three states we will implement are:
+## todos_repository
 
-- `TodosLoadInProgress` - the state while our application is fetching todos from the repository.
-- `TodosLoadSuccess` - the state of our application after the todos have successfully been loaded.
-- `TodosLoadFailure` - the state of our application if the todos were not successfully loaded.
+A [repository](/architecture?id=repository) should specify the interface that the application uses and may provide some other logic.
```

Comment:
I would try to expand a bit more the definiotn give here. A repository is the "business layer": it takes data clients that have no business value, and turns its public API into business-related topics.

For example, a business question would be "I want to be able to load my todos even when I am offline". With this requirement, and knowing that we have a todos_api and local_storage, the repository would simply have a `loadTodos` API that checks both data providers and returns the expected result, without leaking data-acquisition information to the app.

Just saying that a repository `specify the interface that the application uses and may provide some other logic.` can lead to confusion quite quickly.

Reply:
> A repository is the "business layer": it takes data clients that have no business value, and turns its public API into business-related topics.

🙏🏼 this is a very complete definition of business layer that I've been looking for